### ISSUE TITLE
Error Using NVENC For Video Encoding

### ISSUE THREAD

**Issue by morphological** [mpv-player/mpv#5963]
Hi guys!

CPU: i5 6600K
GPU: GTX 970
Kernel: 4.16.0 and 4.8.0 rc1
Driver: nvidia-390 from graphics team ppa
Platform: Kubuntu 16.04.3

**mpv version** (from djcj/vapoursynth repository --> mpv compiled with vapoursynth)

mpv 0.28.0 (C) 2000-2017 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.15.100
   libavcodec      58.19.100
   libavformat     58.13.100
   libswscale      5.2.100
   libavfilter     7.18.100
   libswresample   3.2.100
ffmpeg version: 3.4.git

**Reproduction & Expected Behavior**

` mpv /home/name/Downloads/naruto.mkv -o output --no-audio --no-sub --no-sub-auto --input-ipc-server=mpvencodepipe --ofps 60.000 --of=matroska --ovc=h264_nvenc --ovcopts=b=11055051,rc=vbr_hq,preset=slow,profile=high,threads=16Playing: /home/name/Downloads/naruto.mkv
(+) Video --vid=1 (h264 1920x818 24.000fps)
Audio --aid=1 --alang=eng 'AC3 5.1 @ 640 Kbps' (ac3 6ch 48000Hz)
Using conversion filter.
Using conversion filter.
Image formats incompatible or invalid.
Video filter chain:
[in] 1920x818 yuv420p bt.709/bt.709/bt.1886/limited SP=1.000000 CL=mpeg2/4/h264
[convert] 1920x818 yuv420p bt.709/bt.709/bt.1886/limited SP=1.000000 CL=mpeg2/4/h264 [a]
[out] ??? <---
[convert] ??? [a]
Cannot initialize video filters.
Could not initialize video chain.
Video: no video

Exiting... (Errors when loading file)
[encode-lavc] vo-lavc: encoded 0 bytes
[encode-lavc] ao-lavc: encoded 0 bytes

The graphic's driver works properly and so do any other related packages. I have read that ffmpeg might be the issue as it needs to be compiled with nvenc in order for this to work. Can you confirm this?

**Comment by mc4man** [comment]
That build doesn't have nvdec or nvenc support

**Comment by morphological** [comment]
Thanks for the quick reply.
Could you please explain in detail how I can enable nvenc support?
I do have ffmpeg compiled --enable-nvenc and the necessary headers.

**Comment by mc4man** [comment]
This is not the place for how-to's. You could
1. contact maintainer, ask him to support cuda/nv* via nv-codec-headers
2. Do your own local build
3. Rebuild your ppa's source locally supplying nv-codec-headers yourself
3 may be most straightforward though how-to is better suited to a forum..

**Comment by morphological** [comment]
Yeah, I apologize.
Just one more tiny question if you don't mind.

I first installed the nv-codec-headers and started to compile mpv with ffmpeg option --enable-nvenc ,  --enable-cuvid and mpv option --enable-vapoursynth
The helper script does fine until it reports an error: vapoursynth not found
(despite the fact that I have it installed from via djcj repo)

After some googling I found this mpv issue
https://github.com/mpv-player/mpv/issues/1737

Unfortunately, I couldn't find vapoursynth.pc in /usr/local/lib/pkgconfig or anywhere else using "find"
I then learned that vapoursynth.pc is only generated upon compiling the latest git, which I am now trying to do.

I have installed all dependencies, built zimg, and created symlinks for python, pip and cython to point to their version 3 counterparts, but I still get this error every single time I try to compile.

`Error compiling Cython file:
------------------------------------------------------------
...

    def __repr__(self):
        if self.single:
            return "<Environment (default)>"

        return f"<Environment {self.env_id} ({('active' if self.active else 'alive') if self.alive else 'dead'})>"
               ^
------------------------------------------------------------

`
Could you help me figure this out?

**Comment by mc4man** [comment]
just install the ppa's libvapoursynth-dev package (and undo any weird symlinking, ect.

**Comment by morphological** [comment]
Success! 
I have compiled mpv with vapoursynth, ffmpeg with nvenc/cuvid and transcoding appears to be functioning correctly, with about 50% - 75% Video Engine Utilization being displayed in Nvidia-Settings.
Should anyone else have trouble compiling this on 18.04, forget the git nvenc headers, you need to use the SDK from the Nvidia website, and create and account first....(-.-)

Only a minor issue persists, which I fixed, but I would like to understand the problem behind it.

I had to modify the original command in order for it to work.
(working command) 
`$ mpv /home/name/Downloads/naruto.mkv -o output --no-audio --no-sub --no-sub-auto --input-ipc-server=mpvencodepipe --vf=vapoursynth:/home/name/60fps.vpy:4:16 --of=matroska --ovc=h264_nvenc --ovcopts=b=11055051,rc=vbr_hq,preset=slow,profile=high,threads=16
Playing: /home/name/Downloads/naruto.mkv
 (+) Video --vid=1 (h264 1920x818 24.000fps)
     Audio --aid=1 --alang=eng 'AC3 5.1 @ 640 Kbps' (ac3 6ch 48000Hz)
VO: [lavc] 1920x818 yuv420p
[vo/lavc] Opening encoder: NVIDIA NVENC H.264 encoder [h264_nvenc]
[encode] Opening output file: output
[encode] Opening muxer: Matroska [matroska]
V: 00:04:36 / 00:14:48 (31%) {5.5min 111.6fps 2349.6MB}
V: 00:04:42 / 00:14:48 (31%) {5.4min 111.6fps 2354.5MB}
[input] No key binding found for key 'Alt+q'.
V: 00:04:55 / 00:14:48 (33%) {5.3min 111.3fps 2370.3MB}`

When I  used this  option, an error occured.
`--ofps 60.000`
The suggested solution was to use --vf-add instead.

Has the option been deprecated, considering that I am using the latest git of mpv?

Thank you for your help, you guys are really doing an amazing job here.
Every single time I have presented an issue here, it has been solved. 

Please leave this issue open for another day, as I am going to write a complete and detailed guide on how to enable nvenc transcoding with mpv on Ubuntu 18.04 and would like to post the link here.

**Comment by morphological** [comment]
anyone?

**Comment by mc4man** [comment]
I think the message tells you all you'd need, the option has been removed, suggest an option to try. Whether that option suits your purpose vs. using a vs script you'd have to test & see.

In regards to building with the nv-codec-headers, see no issue here. For 18.04 I decided there is some value to vs support, put a build here. Have no real interest in supplanting the vs ppa, this was just for 18.04 & provides  a bit more complete package than the vs ppa one
https://launchpad.net/~mc3man/+archive/ubuntu/bionic-media
